controllers/storagecluster/volumegroupsnapshotterclasses.go

```
@@ -0,0 +1,175 @@
+	if !r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
+		r.Log.Info("VolumeGroupSnapshotClass CRD is not available")
+		return reconcile.Result{}, nil
+	}
+
+	vgsc := newGroupSnapshotClassConfigurations(instance)
+
+	err := r.createGroupSnapshotClasses(vgsc)
+	if err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (obj *ocsGroupSnapshotClass) ensureDeleted(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+	if !r.AvailableCrds[VolumeGroupSnapshotClassCrdName] {
+		r.Log.Info("VolumeGroupSnapshotClass CRD doesn't exist")
+		return reconcile.Result{}, nil
+	}
+
+	vgscs := newGroupSnapshotClassConfigurations(instance)
+	for _, vgsc := range vgscs {
+		sc := vgsc.groupSnapshotClass
+		existing := groupsnapapi.VolumeGroupSnapshotClass{}
+		err := r.Client.Get(r.ctx, types.NamespacedName{Name: sc.Name, Namespace: sc.Namespace}, &existing)
+
+		switch {
+		case err == nil:
+			if existing.DeletionTimestamp != nil {
+				r.Log.Info("Uninstall: GroupSnapshotClass is already marked for deletion.", "GroupSnapshotClass", klog.KRef("", existing.Name))
+				break
+			}
+
+			r.Log.Info("Uninstall: Deleting GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef("", existing.Name))
+			existing.ObjectMeta.OwnerReferences = sc.ObjectMeta.OwnerReferences
+			sc.ObjectMeta = existing.ObjectMeta
+
+			err = r.Client.Delete(r.ctx, sc)
+			if err != nil && !errors.IsNotFound(err) {
+				r.Log.Error(err, "Uninstall: Error deleting the GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef("", existing.Name))
+				return reconcile.Result{}, err
+			}
+		case errors.IsNotFound(err):
+			r.Log.Info("Uninstall: GroupSnapshotClass not found, nothing to do.", "GroupSnapshotClass", klog.KRef("", sc.Name))
+		default:
+			r.Log.Error(err, "Uninstall: Error while getting GroupSnapshotClass.", "GroupSnapshotClass", klog.KRef("", sc.Name))
```

dont we need to return `err` if something fails? same problem exists in storageclass and volumesnapshot class as well. @iamniting can you please confirm this?

---

i think in that ase if will be `errors.IsNotFound` not any other error, this could leave us to a case where we might leave the stale classes due to some internal API server errors or something else. i will let @iamniting to confirm it, except this one everything else LGTM

---

If there is an error we should return an error IMO.